### PR TITLE
WIP: allow add_srcdata to place sources on non-owned points

### PR DESCRIPTION
**Pull request by mawc2019**
This PR continues @stevengj 's work in #1547. At present, it still fails when too many processes are involved, which can be illustrated by the [example](https://github.com/NanoComp/meep/pull/1547#issuecomment-963528760) with 7 processes. The error is `unknown src_time_id (missing registration?)`.

### CODE REVIEW

File: src/fix_boundary_sources.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        }
+    }
+  }
+
+  // we need each process's data to be contiguous
+  srcpt_info_compare compare = {chunks};
+  std::sort(boundarysources.begin(), boundarysources.end(), compare);
+
+  // collect 2d (row-major) arrays offsets and numcomm,
+  // where numcomm[i,j] is the number of srcpt_info items
```
Comment (stevengj) [inline src/fix_boundary_sources.cpp:84]: Note: `numcomm[i,j]` here is stored at `numcomm[i*P + j]` (this is "row major order").

**Comment by mawc2019** [comment]
In `fix_boundary_sources.cpp`, although the value of `srcpts[0].src_time_id` in the source process immediately after `MPI_Send(srcpts, N, mpi_srcpt_info, pdest, psrc*P + pdest, mycomm)` at Line [164](https://github.com/NanoComp/meep/blob/997a1a720425f7b98c33135ca0d023037521dd01/src/fix_boundary_sources.cpp#L164) is the same as that in the destination process immediately after `MPI_Recv(srcpts, N, mpi_srcpt_info, psrc, psrc*P + pdest, mycomm, &status)` at Line [129](https://github.com/NanoComp/meep/blob/997a1a720425f7b98c33135ca0d023037521dd01/src/fix_boundary_sources.cpp#L129), `src_time *srctime = lookup_src_time(srcpts[0].src_time_id)` gives different `srctime`, which is not `NULL` in the source process but is often `NULL` in the destination process. The null `srctime` in the destination process causes the [error](https://github.com/NanoComp/meep/blob/997a1a720425f7b98c33135ca0d023037521dd01/src/fix_boundary_sources.cpp#L148) later. Occasionally the error does not appear even though the same code is run.

**Comment by stevengj** [comment]
Went over the code with @mawc2019 today, and found a typo in the `srcpt_info_offsets` definition which was causing the problem noted [above](https://github.com/NanoComp/meep/pull/1959#issuecomment-1062370494).

**Comment by mawc2019** [comment]
The error message of the online check is as follows, but that c++ file already has `#include <algorithm>`.
```
../../../src/fix_boundary_sources.cpp: In member function ‘void meep::fields::fix_boundary_sources()’:
../../../src/fix_boundary_sources.cpp:81:8: error: ‘sort’ is not a member of ‘std’; did you mean ‘sqrt’?
   81 |   std::sort(boundarysources.begin(), boundarysources.end(), compare);
      |        ^~~~
```
There is no error when it is checked offline.

File: src/fix_boundary_sources.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+#ifdef NEED_UNDEF_SEEK_FOR_MPI
+// undef'ing SEEK_* is needed for MPICH, possibly other MPI versions
+#undef SEEK_SET
+#undef SEEK_END
+#undef SEEK_CUR
+#endif
+#include <mpi.h>
+#include <algorithm>
+#endif
+
```
Comment (stevengj) [inline src/fix_boundary_sources.cpp:14]: You are only doing `#include <algorithm>` on MPI builds (it is inside the `#ifdef HAVE_MPI`) — you have to do it on serial builds too.

```suggestion
#include <mpi.h>
#endif

#include <algorithm>

```

**Comment by stevengj** [comment]
See my suggested patch above.

File: python/adjoint/objective.py [hunk 3]
```diff
@@ -273,12 +273,12 @@ def place_adjoint_source(self, dJ):
             scale = amp_arr * self._adj_src_scale(include_resolution=False)
             
             if self.num_freq == 1:
-                sources += [mp.IndexedSource(time_src, fourier_data, scale[:,0])]
+                sources += [mp.IndexedSource(time_src, fourier_data, scale[:,0], not self.yee_grid)]
```
Comment (stevengj) [inline python/adjoint/objective.py:276]: Note that this only affects code that use Fourier fields `yee_grid=false`.

**Comment by mawc2019** [comment]
Here is an example. Without this PR, the finite-difference and adjoint gradients are inconsistent when the number of processes is large, such as 8. With this PR, the finite-difference and adjoint gradients become consistent.
image
The code is as follows.

```
import meep as mp
import meep.adjoint as mpa
import numpy as np
from autograd import numpy as npa
Si, SiO2 = mp.Medium(index=3.4), mp.Medium(index=1.44)

resolution,design_resolution = 20,20
Lx,Ly,pml_size = 2,4,1
cell_size = mp.Vector3(Lx,Ly+2*pml_size)
pml_layers = [mp.PML(pml_size, direction=mp.Y)]
design_width,design_height = Lx-0.4,0.4

fcen,fwidth = 1/1.55,0.01/1.55
source_center,source_size = mp.Vector3(0,-Ly/4),mp.Vector3(Lx,0)

src = mp.GaussianSource(frequency=fcen,fwidth=fwidth)
source = [mp.Source(src,component=mp.Ez,center=source_center,size=source_size)]
Nx,Ny = int(round(design_resolution*design_width)),int(round(design_resolution*design_height))

design_variables = mp.MaterialGrid(mp.Vector3(Nx,Ny),SiO2,Si,grid_type='U_MEAN')
design_region = mpa.DesignRegion(design_variables,volume=mp.Volume(center=mp.Vector3(), size=mp.Vector3(design_width, design_height)))

geometry = [mp.Block(center=design_region.center, size=design_region.size, material=design_variables)] # design region
sim = mp.Simulation(
    cell_size=cell_size,boundary_layers=pml_layers,geometry=geometry,
    sources=source,default_material=SiO2,resolution=resolution)

nf = mpa.FourierFields(sim,mp.Volume(center=mp.Vector3(0,design_height+0.2),size=mp.Vector3(Lx,0)),mp.Ez)
ob_list = [nf]

def J(near_data):
    return npa.sum(npa.abs(near_data)**2)

opt = mpa.OptimizationProblem(
    simulation = sim,objective_functions = J,objective_arguments = ob_list,design_regions = [design_region],
    fcen=fcen,df = 0,nf = 1,decay_by = 1e-5)

x0 = 0.5*np.ones(Nx*Ny)
opt.update_design([x0])
f0, dJ_deps = opt()

db, choose = 1e-4, 5
np.random.seed(20211204)

g_discrete, idx = opt.calculate_fd_gradient(num_gradients=choose,db=db)
g_discrete = np.array(g_discrete).flatten()
g_adjoint = dJ_deps.flatten()

print("Randomly selected indices: ",idx)
print("Finite-difference gradient: ",g_discrete)
print("Adjoint gradient: ",g_adjoint[idx])
```

**Pull request merged**